### DISCUSSION TITLE
Use Jotai in Next.js app router

### DISCUSSION THREAD

**Discussion by ootkin** [pmndrs/jotai#2337] [Q&A]
Hi,

I would like to integrate Jotai in a next.js (v14) using app routers.

Following the [Jotai documentation](https://jotai.org/docs/guides/nextjs) (even if the example reports page routers instead) I see that we need to use a global Provider.

I tried to implement the Provider in the RootLayout (`/app/layout.tsx`) but it seems like to work only when i refresh the page: when i change page using `router.push()` the state is not hydrated with `useHydrateAtoms`.

Do you have any ideas? Maybe the Provider must be placed in each page and in the root layout.

As the [Next.js documentation](https://nextjs.org/docs/app/building-your-application/routing/pages-and-layouts#layouts) says, the layout persist across routes and maintain state.

Thanks

**Comment by dai-shi** [comment]
Basically, the route change is not hydration. Hydration is only for the initial render.

So, we need to update with both useHydrateAtoms and useEffect for the route change.

...if I'm not mistaken.

The idea is this:
```js
const useSyncAtom = (atom, value) => {
  useHydrateAtoms([atom, value]);
  const setAtom = useSetAtom(atom);
  useEffect(() => {
    setAtom(value);
  }, [setAtom, value]);
};
```

**Comment by ootkin** [comment]
Well, when i change navigation to a new Page (wich is a Server Components) i fetch some data and return them as a prop to the Page.

So a new state is computed server side and I think it is an hydration.

Right now I'm looking [this](https://jotai.org/docs/guides/initialize-atom-on-render#using-typescript)

**Reply by ghost** [reply]
I get this error 
```
Error: Functions cannot be passed directly to Client Components unless you explicitly expose it by marking it with "use server". Or maybe you meant to call this function rather than return it.
  {toString: ..., init: "", read: ..., write: function defaultWrite}
```
I use nextjs 14 and approuter

**Comment by ootkin** [comment]
Furthermore, if I use `useHydrateAtoms` within the Jotai Provider, I've got this error:

Detected multiple Jotai instances. It may cause unexpected behavior with the default store. https://github.com/pmndrs/jotai/discussions/2044

**Reply by dai-shi** [reply]
If you see the error without HMR, it sounds bad. You are using Provider right? You may want to use Provider separately for different routes.

**Reply by ootkin** [reply]
I use HMR and I use Provider separately for different routes: even after I removed useHydrateAtoms.

**Comment by MiroslavPetrik** [comment]
@ootkin it's a feature of the `useHydrateAtoms` to allow only one hydration.

Let's say you have a page like `app/[userId]/page.tsx` with a jotai provider in the `app/layout.tsx`.
When you start by navigating to `/1`, the hydration will work.
Next, if you `router.push("/2")` the **hydration won't work second time for the same atom.**

This might be confusing, as you are correct - the SSR takes place, so we expect hydration... but since the [`router.push` is client side navigation](https://nextjs.org/docs/app/api-reference/functions/use-router#userouter), we must keep in mind that _the app has a local state in the memory._. Looking into the `useHydrateAtoms` source, we can see that there is a [local hydratedSet created for each store.](https://github.com/pmndrs/jotai/blob/2d760920f5ebea05abaa8e072cee578fa125ca9b/src/react/utils/useHydrateAtoms.ts#L54).

This means, that if you have `<Provider>` in your `RootLayout`, there is a shared store for all of your routes. And so the atom hydration works  **per app**, not _per route_ because the `hydratedSet` stays the same...

**So the solution is to use the `<Provider>` in your `page.tsx`.**

Other (hypothetical) solution could be to call `hydratedSet.delete` for all the atoms you wish to rehydrated, just before the `router.push`.
That would require changes to the util, as the hydratedSet is currently not accessible... perhaps the hook could return `evictAtoms` cleanup function.
```ts
const {evict} = useHydrateAtoms([[atom, value]]);
```

This could help for other client side navigation cases like `router.refresh()` which is usually called after a form submit.
Here, we need to reinitialize/rehydrate the form, which won't happen despite the page having a local `<Provider>`.
So invalidating the cache (the `hydratedSet` here) of already hydrated atoms could help.

```ts
evict() // enables subsequent hydration
router.refresh() // fetches new data & merges the RSC
// form will be reinitialized 
```

With such cache eviction we won't need to set the atoms manually after the hydration
 ```ts
useEffect(() => {
    setAtom(value);
  }, [setAtom, value]);
```
This to me looks like a "second hydration" and can be confusing, or easy to forget. 

@dai-shi thoughts?

**Reply by MiroslavPetrik** [reply]
I've just tried the `evict` idea together with the `router.refresh` and as it turns out it causes react render error, that setState is called while another rendering is taking place. So yeah, I was able to clear the hydratedSet, which should permit the regular hydration done by `useHydrateAtom` via `store.set(atom, value)` so that call is likely not permited together with `router.refresh()` makes me wonder if the 'merging of RSC' which happens is different to regular render and how. 

Also this makes me wonder what the `dangerouslyForceHydrate` option is for. 

```ts
// this would be necessary to export from `jotai/utils/useHydrateAtoms` 
import { getHydratedSet } from "./useHydrateAtoms";

export const useRehydrateAtoms = (atoms: readonly AnyWritableAtom[]) => {
  const store = useStore();
  const hydratedSet = getHydratedSet(store);

  return useCallback(() => {
    for (const atom of atoms) {
      hydratedSet.delete(atom);
    }
  }, atoms);
};
```

This is irrelevant to the original issue, just my exploration of the jotai hydrate in next.js

**Reply by dai-shi** [reply]
This has been raised a few times, and we haven't decided to what to recommend, but our general idea is to use `useEffect` for the second hydration.

I'm not sure if "rehydrate" is the right word, but the code is something like this:
```js
const useRehydrateAtoms = (atomValues) => {
  useHydrateAtoms(atomValues);
  const rehydrate = useSetAtom(useMemo(() => atom(null, (_get, set) => {
    for (const [a, v] of atomValues) {
      set(a, v);
    }
  }), [atomValues]));
  useEffect(() => rehydrate(), [rehydrate]);
};
```

We could use `useStore` instead.

**Reply by michaelschufi** [reply]
It seems that we need to wrap the arguments into a `useMemo` as well. 

Doing this
```ts
useRehydrateAtoms([[countAtom, countFromServer]])
```
like this
```ts
useHydrateAtoms([[countAtom, countFromServer]])
```
won't work with the above hook. Since the tuple will get a new reference in each render. What needs to be done is this:

```ts
useRehydrateAtoms(
  useMemo(
    () =>
      [[countAtom, countFromServer]],
    [countFromServer],
  ),
);
```
Correct?

**Reply by dai-shi** [reply]
Yes, nice catch!

**Reply by Tomaszal** [reply]
At that point wouldn't it be easier to simply use `useMemo` to both hydrate and rehydrate the atom? 

```ts
const store = useStore();
useMemo(
  () => store.set(countAtom, countFromServer),
  [store, countFromServer],
);
```

I realize this isn't really a semantically correct use of `useMemo`, but it seems to solve this issue pretty well. It's executed immediately before render (as opposed to `useEffect`), and subsequently when "rehydration" is needed for RSC. Am I missing some footgun here?

**Reply by Tomaszal** [reply]
A wrapper for this would also be really simple and would hide the non-semantic use of `useMemo`:

```ts
const useSyncAtoms = (atomValues) => {
  const store = useStore();
  for (const [atom, value] of atomValues) {
    useMemo(() => store.set(atom, value), [store, atom, value]);
  }
};
```

**Reply by dai-shi** [reply]
That breaks the rules of React. No side effects in render. `useHydrateAtoms` uses an exception of the rules. https://gist.github.com/sebmarkbage/75f0838967cd003cd7f9ab938eb1958f

**Reply by Tomaszal** [reply]
What exception does `useHydrateAtoms` fall under?

I guess that wrapper can still be used with `useHydrateAtoms` and `useEffect` instead like so:

```ts
const useSyncAtoms = (atomValues) => {
  useHydrateAtoms(atomValues);
  const store = useStore();
  for (const [atom, value] of atomValues) {
    useEffect(() => store.set(atom, value), [store, atom, value]);
  }
};
```

I think it should have the same effect as the snippet provided by you earlier (https://github.com/pmndrs/jotai/discussions/2337#discussioncomment-8151007), but without having to separately memoize the inputs, as their references get passed to `useEffect` directly. Or is there a reason you were creating a new write-only `rehydrate` atom?

**Reply by dai-shi** [reply]
useEffect in a for-loop can violate the rules of hooks. The length of `atomValues` can vary.

The trick in useHydrateAtoms is that it's only effective for uninitialized atoms.

**Reply by Tomaszal** [reply]
Right, for some reason I thought it wouldn't be a problem since `useSyncAtoms` is a hook itself, but yeah of course the array can still change.. Still I think there's a way to do it without an extra atom and memoization like so (unless I'm missing something obvious again):

```ts
const useSyncAtoms = (atomValues) => {
  useHydrateAtoms(atomValues);
  const store = useStore();
  useEffect(
    () => atomValues.forEach(([atom, value]) => store.set(atom, value)),
    [store, ...atomValues.flat()],
  );
};
```

**Reply by dai-shi** [reply]
This is not a Jotai issue, but a React issue. You can't safely do `useEffect(fn, [...arr])` or just `useEffect(fn, arr)` unless `arr` is stable. If the length of `arr` changes, React will warn you in dev mode.

**Reply by Tomaszal** [reply]
I realize this isn't a Jotai issue, wasn't implying that. I'm just looking for a way to do this in a wrapper that doesn't require memoizing each atom's value before passing it to `useRehydrateAtoms`. I guess the simplest way would be to use a wrapper for individual atoms (as in https://github.com/pmndrs/jotai/discussions/2337#discussioncomment-8092610) instead of specifying an array of tuples, I was just trying to keep the API similar to the existing `useHydrateAtoms`, but seems like there isn't a clean way to do that.

**Reply by dai-shi** [reply]
Dealing with arrays in React is always a bit tricky.

**Reply by jonahallibone** [reply]
Has anyone created a typed verison of this hook? Struggling to figure this out with the types that are exposed.

**Comment by dai-shi** [comment]
related: #669

**Comment by benjamin-andersen** [comment]
> https://github.com/pmndrs/jotai/discussions/2337#discussioncomment-8142862

@ootkin I believe you can use the root template for the app router in Nextjs. It should re-create the `<Provider>` anytime a user navigates between routes. This will allow the hydration to occur every time the route changes, and has the added benefit of not needing to wrap every page in the `<Provider>`.
https://nextjs.org/docs/app/building-your-application/routing/layouts-and-templates#templates

I did some rudimentary testing with this and it seemed to fix the issue for me.
```tsx
// app/template.tsx

import {Provider} from 'jotai'

interface RootTemplateProps {
    children: React.ReactNode
}

const RootTemplate: React.FC<RootTemplateProps> = ({children}) => (
    <Provider>
        {children}
    </Provider>
)

export default RootTemplate
```

If you want to synchronize the hydration with cache invalidation (e.g. `revalidateTag()`), then you will need to use something like [`this hook`](https://github.com/pmndrs/jotai/discussions/2337#discussioncomment-8210660) instead of `useHydrateAtoms()`, since the hydration occurs only once per route change.
